### ISSUE TITLE
long options 2023 restart

### ISSUE THREAD

**Issue by rbdavis** [PaulWessel/CodingTasks#6]
Starting new issue for ongoing discussion of longopts mods beginning June 2023.

**Comment by rbdavis** [comment]
Hi Paul,

> I see -D in docs, synopsis, and usage message, so -D is real.

Sorry, but no luck finding any **batch -D** option documentation. I'm looking at the [latest man page](https://docs.generic-mapping-tools.org/latest/batch.html) (I think) but see no -D there, nor any -D on the common options page.

Beginning work on other **batch** longopts now. I don't see any existing test which exercises **batch**, if you are aware please advise. If there's nothing maybe I'll try to use the **batch** example given at the bottom of the above man page, but if there's any other quick test available for use that might save time.

**Comment by PaulWessel** [comment]
You are looking at 6.4.0 docs but need to select dev version (top left menu of versions)

**Comment by rbdavis** [comment]
Got it, thanks!

**Comment by rbdavis** [comment]
Built and ran the second example (printing the country maps). Worked fine, nice feature -- very surprised to see that it deleted all of the scripts I created, though! Quickly figured out this was due to the **-Z** option, but that option is actually invoked within the example as **-Zs**, and I can't find any description of an **s** modifier for **-Z**.  ??

**Comment by rbdavis** [comment]
Any reason not to longoptify the **begin** module? You did not create a **longopt/begin_inc.h** file, but it does seem to have the same kind of module startup code as other modules (e.g., compare **begin.c** with **batch.c** ). Its **-C** option probably deserves a **--clean** longopt variant?

**Comment by PaulWessel** [comment]
Not really now that it has an option (-C).  Yes, lets do the --clean.

**Comment by rbdavis** [comment]
Will do (re: begin --clean). FYI, I'm trying to work full days on each of my two projects wherever possible. I did 3 days on GMT, 2 on ARL this past week, so will invert next week. Likely next GMT day will be mid-week.

**Comment by rbdavis** [comment]
Working on **dimfilter** now. Your placeholder **dimfilter_inc.h** has entries for **-C**, **-E** and **-S**, but the man page (dev version) does not mention these?

Also, it looks like **dimfilter**'s **-I** is different than the quasi-global **GMT_INCREMENT_KW**, so I'm not using the latter in the table.

**Comment by rbdavis** [comment]
For **dimfilter**'s **-D** I used **--gfdistance** because **--distance** is already used in the common options for **-j** (and is found first in the entry search). Would you prefer anything else?

**Comment by PaulWessel** [comment]
For -D I think we don't want the word distance involved since it is more of a mode setter. Try --disttype.  Unfortunately I do not see Julia version of dim filter nor grdfilter, but those should have many similar options.  You might want to do grdfilter after dim filter or at the same time.
Not sure about the place holder - might be junk..
No, -I is the same (grid spacing).  It just developed as a supplement so the docs are trailing.  In fact, you should compare with grdfilter (as well as grdfilter.rst) and make dim filter report the same for -I as grdfilter does.
-F and -N might become primary-filter and secondary-filter perhaps.

**Comment by rbdavis** [comment]
OK, so you're saying the **dimfilter** man page is incorrect and I should just use the standard **GMT_INCREMENT_KW** entry ?

**Comment by PaulWessel** [comment]
Yep

**Comment by rbdavis** [comment]
Working now on **grdfilter** together with **dimfilter** as you suggested.

**dimfilter** man page is really confusing. The **-F** option summary says -Fxwidth[+l|**u**], but then the description immediately below that says

_(p) Maximum likelihood probability (a mode estimator): Return modal value. If more than one mode is found we return their average value. Append +l or_ **+h** _to the filter width if you rather want to return the smallest or largest of each sector’s modal values._

So, is it **+u** (from the summary) or **+h** (from the description)?

The same question also applies to **dimfilter**'s **-N** option, where again the summary says **+u** and the description says **+h**.

I see that **grdfilter**'s **-F** option is more consistent, referring to **-u** in both its summary and description (although it also has its own **+h** modifier which means something entirely different (high-pass).

**Comment by PaulWessel** [comment]
Sorry, in bed. Just check dimfilter.c parse function to see if +h +u or both. I know this is messy. Code hacked from GMT 4 by Seung-Sep 18 years ago then changed over time to a module…

**Comment by rbdavis** [comment]
OK, have a good night's sleep! There may be more followups on this but I hope you will turn your phone off and get some rest!

**Comment by rbdavis** [comment]
Been wrestling with **grdfilter -G** checking to see if obscure GDAL stuff will work, but I'm having trouble finding a combination of GDAL parameters that work together. (All I want to see is if the parsing code works -- I believe that it is, but it would be nice to get a stronger confirmation with no error messages whatsoever and a verifiable output product. Help! ;->

For example, here's a shortopts command that is legal, I think:

gmt grdfilter AFR.nc -Fg100+h -D2 -Gxx.jpg=gd:JP2OpenJPEG/i32+cQUALITY=100+cREVERSIBLE=YES+cYCBCR420=NO

When I longoptify this I get:

gmt grdfilter AFR.nc --filter=gaussian:=100+highpass --disttype=2 \
   --outgrid=xx.jpg=gd:JP2OpenJPEG/u8+gdal=QUALITY=100+gdal=REVERSIBLE=YES+gdal=YCBCR420=NO

When I run the above longoptified command it appears to be correctly parsed but GDAL complains about driver issues:

```
>  gmt grdfilter AFR.nc --filter=gaussian:=100+highpass --disttype=2 \
   --outgrid=xx.jpg=gd:JP2OpenJPEG/u8+gdal=QUALITY=100+gdal=REVERSIBLE=YES+gdal=YCBCR420=NO
gmt [WARNING]: Reformatted options: AFR.nc -Fg100+h -D2
   -Gxx.jpg=gd:JP2OpenJPEG/u8+cQUALITY=100+cREVERSIBLE=YES+cYCBCR420=NO
grdfilter [ERROR]: gmt_gdalwrite: The Jpeg2000 driver does not support floats.
```

This sort-of seems like 'good enough' to me insofar as the longopt parsing goes. (Note that the above line which begins with  '**gmt [WARNING]: Reformatted options**:' appears to be a correct reformatting.) If you agree I will call it done and submit the PR when I return to work on this next week, if not then I guess I need to have someone give me a functional grdfilter GDAL output command that calls multiple +c GDAL options.

**Comment by PaulWessel** [comment]
I am consulting with Joaquim how to do those crazy GDAL extensions.  My gut says leave as it was for now and just do --outgrid since -G is used extensively in GMT.

**Comment by rbdavis** [comment]
It does appear from my testing that the longoptions parser is correctly translating my **+gdal** longopt modifier to the old **+c** shortopt modifier, so I think things are very likely OK. If someone can give me a test command that uses the proper combination of GDAL driver and datatype with preferably multiple **+c** modifiers which are compatible with that driver (which is the unknown to me) that would be the final proof, I think.

**Comment by rbdavis** [comment]
Just submitted PR for {dim,grd}filter. If you think the +gdal modifier in grdfilter is too sketchy please go ahead and remove it!

**Comment by rbdavis** [comment]
Finished re-testing of all existing longoptified modules (no issues!), will start looking at longopt-translator parser test next week. Note the still-outstanding PR for dimfilter/grdfilter that was submitted pre-Oslo.

**Comment by PaulWessel** [comment]
Will try to have a look this weekend. Family has forced a 5-day road trip  to Ålesund, but bringing laptop!

cheers Paul

**Comment by rbdavis** [comment]
No worries. I'll start poking around the general GMT option parser next week to figure out the best way to add the proposed --dumptranslation  option, and as soon as the PR on the two outstanding filter modules is cleared I'll upload the code for the updated l2s translation of the : arguments.

**Comment by rbdavis** [comment]
Hi Paul, 

Hope things are going well over there these days!

I have implemented a **--l2stranstest** option to trigger the requested long-to-short translation test and then immediately exit. Here is some sample output:

```
> gmt psxy -R1/2/3/4 -V --l2stranstest
L2STransTest_NoTranslation

> gmt psxy --region=1/2/3/4 --verbosity --yabbadabbadoo --l2stranstest
L2STransTest_-R1/2/3/4 -V --yabbadabbadoo --l2stranstest
```

It's probably worth discussing the (trivial!) implementation of this a bit. I spent the lion's share of the day trying with limited success to understand three aspects of the gmt_init and gmt_api routines, namely (i) general option-parsing, (ii) error-reporting and general messaging, and (iii) process shutdown code, all of which seem to be pretty involved. In the end I opted for what's probably the simplest (but not necessarily fully correct) solution as follows:

(1) I decided to just test for the special **--l2stranstest** option within the main long-to-short translation routine rather than in the general option-parsing code, which is quite complex. This turned out to be very easy to do, and also has the virtue of segregating the option within the code block that's purely about longopts translation itself. I'm quite confident this is the right way to go on this point.

(2) As far as printing out the translation result and exiting, I basically just directly used **fprintf()** and **exit()**. I'm less sure this is the correct thing to do, but it certainly works for my testing purposes. I spent a long time trying to puzzle through the details of **GMT_Report()** and **GMT_Message()**, either of which might be better choices than **fprintf()**, but could not figure out the best way to print a message that was absolutely guaranteed to be written to **stdout** with a minimal amount (preferably none!) of unrelated chatter. Similarly, I also looked at **GMT_Destroy_Session()** as an alternative to **exit()**, but again could not really figure out if this was a necessary or even desirable thing to do.

So, this is where I could use some advice. Right now the code block very near the end of the translation routine is this:

```
if (transtest) {
   if (modified) {
        char *cmd = GMT_Create_Cmd (API, *options);
        (void) fprintf(stdout, "L2STransTest_%s\n", cmd);
        GMT_Destroy_Cmd (API, &cmd);    /* Free string */
    }
    else
        (void) fprintf(stdout, "L2STransTest_NoTranslation\n");
    (void) fflush(stdout);
    exit (0);
}
```

Is this OK or should we be using GMT functions for message output and/or shutdown, and if so please specify in detail.

Thanks!

PS: Still waiting on PR merge for the dimfilter/grdfilter changes.

**Comment by PaulWessel** [comment]
Hi Roger, I know we agreed on a weeklong review for pull requests but people are still partly on travel (as have I) and now NSF wants two annual reports from me and working with Lily on creating a RCUH job for me so I can receive some salary for working (!).  I also need to turn our rough meeting notes into a report.  So, hopefully next week will be on where I get this going, including the filter PR and check your tests.

orubtf/forintf is fine for this - no need to dip into the general stuff for this since it will not (likely) be a user accessible thing,

**Comment by rbdavis** [comment]
Yikes, you should definitely be getting paid for this! Lily's been very efficient at handling my travel stuff, my reimbursement check is supposedly already in the mail -- hopefully she can get your RCUH job set up quickly.

If you're OK with my quick-and-dirty print and exit then the code changes to the translation routine for the : argument processing and the translation testing are all done. I will start working on the actual translation test scripts now, and of course there's a decent-sized backlog of those to do for all of the modules which have already been long-optified. I might have some questions for you about general test script structure if I can't puzzle things out on my own. For now I suppose I'll just keep all the translation tests in the same test directory, if you have any thoughts on that please advise.

Just a reminder, I will be on leave from 8/20 through the end of September (mainland travel). So, two more weeks, i.e., 5 working days on GMT (plus today), until that leave begins.

**Comment by PaulWessel** [comment]
Will ponder the testing structure, and your vacation is now in my calendar.

**Comment by rbdavis** [comment]
Two questions:

(1) Any thoughts on how thorough these translation tests should be for each module? We have all the common options (--region, --projection, etc.) plus all the module-specific options, all with potentially multiple directives, modifiers, etc. If you want to test every possible combination of all these things each test script could be huge. I think that would be overkill. Do we need to test all common longopts with every single module? Do we need to test every module-specific long directive and long modifier? For now I will start working on modules in roughly the same order in which I long-optified them (beginning with surface) and be extremely thorough. Please advise if you want me to back off on that to any degree. I'll send you the first one when done to see what you think.

(2) With regard to test scripts in general, I'm not seeing any way to indicate success/failure. I looked at test/README.tests which says

```
# The script should report a PASS/FAIL grade so
# that it is easy to tell which test failed.
```

but I am not seeing any existing test script, e.g., test/mapproject/azimuth.sh, which actually does this. The scripts seem to just all be coded to run some collection of gmt modules without explicitly reporting any result at all. ???

**Comment by PaulWessel** [comment]
1. Let’s see what that surface test ends up looking like. Since we are not parsing and running surface one could test lots of items at once even if in real life some option combinations would yield an error without —l2stranstest
2. Most table-producing tests (e.g., test/block mean) produces a results file which diff compares to a known answer file and writes differences to a file called fail. If fail is empty then the test passed. The gmtest script that runs the tests handles both text and Postscript results.

**Comment by rbdavis** [comment]
OK, just looked at test/gmtest.in and got a bit more insight into this process, but I'm still well in the dark. Essentially, I'm not seeing how text file comparisons are done, which is what will be required here I think. I see lots of stuff that relates to PS files, and poking around test/baseline it seems like everything there is a PS file. I'm sure I'm just being stupid, but I expect for the moment my time is better spent just ginning up a dumb script that spits out a bunch of translated command lines to stdout. Presumably we can run that once to generate the 'correct' translations, archive that output somewhere (in test/baseline?) and then when the test is run the stdout from the test script can be somehow compared with the known-to-be-correct output we archived earlier. If I'm totally barking up the wrong tree please advise (after you get a night's sleep, no need to rush a reply! ;-> ).

Thanks!

**Comment by PaulWessel** [comment]
See line 48 in test/blockmean/io.sh.  If the file _fail_ is empty then we pass else report error.  So for small files I typically use 

```
cat << EOF > answer.txt
correct output goes here
EOF
```

and then the script runs and also creates results.txt (i.e., your output) and then we diff those two.

**Comment by rbdavis** [comment]
Thanks, will follow that model!

I'm working for a while today as yesterday's working hours took a substantial hit due to my MacbookPro dying for the zillionth time. I don't use that for the GMT work but really want to get it fixed (must be shipped to the mainland and back) before I leave for vacation, so another fun time at the Apple Store berating yet another hapless Genius. Glad I'm on the edge of retirement, I guess, as I don't know what I'd buy for my next work laptop. Been using Macs for 15 years but their latest products are absolute shit for reliability, unfortunately I don't see anything else around that I have any reason to believe would be better. Hope my new personal Mac Studio lasts longer, I spent my own $$ on that. Don't skimp on AppleCare!!

**Comment by PaulWessel** [comment]
Yikes! I’ve had good luck with Apple over 20 years. Replaced the big motherboard on a 2007 Mac Pro once (on warranty) but all my 6+ laptops have never had hardware issues.

**Comment by rbdavis** [comment]
Never had any trouble with desktops, but their recent laptops? Grrrr.... I was doing great until the last purchase around 2016. This was the infamous butterfly keyboard model (first one with the touchbar). Multiple failures on keys, each one of which required a roundtrip to the mainland, a brand new mainboard (keyboard could not be separately repaired), and a full system reinstall (SSD also part of mainboard). After the third such failure they finally replaced it with a newer MBP model with the resurrected non-butterfly keyboard. Thought my problems were over -- boy was I an idiot. The new one has a great keyboard but disastrous power supply issues. This is now the second time it has gone dead-as-a-doornail with no prior hint of trouble. Also, I had multiple intermittent episodes of video issues (weird flashing green bands, etc.), and during one of its mainland journeys for that problem some shitclown at their repair facility couldn't duplicate the issue so they just sent it back after thoughtfully wiping the drive and reinstalling macOS. (Another full restore from backup for me for absolutely no reason, thanks a bunch!) I had to catch it in the act and make a video of it before they finally agreed there was a problem and fixed it.

**Comment by rbdavis** [comment]
With regard to multi-platform constraints on the test scripts, is it acceptable to use common Unix commands (e.g., cp, mv, ls, sed, echo, awk, etc.) within those? I'm certainly not planning to do anything grand here in that regard, just generally curious.

**Comment by PaulWessel** [comment]
Yes, full *nix is fine.

**Comment by rbdavis** [comment]
OK, done with the surface test script, will e-mail to you separately. Wasn't as much of a mess (or as large) as I was expecting, but you may think we need to do more testing of various combinations. I made some further changes to gmt_init.c to simplify the --l2stranstest output now that I have a clearer picture on its usage.

I did the following in an attempt to simplify and also streamline creation of tests for remaining modules:

(1) Testing broken out into three sub-groups of common, quasi-common and module-specific options. Makes reasonable logical sense and allows for quick  cut/paste of common options from one test to the next. (Also employed a $module variable to facilitate such copying.)

(2) Each command tests only one longopt (plus aliases), in many cases doing multiple translations of the same longopt within one command.

(3) Each longopt string (primary longopt specifier, each longopt directive and modifier, and each alias for any of the foregoing) is tested at least once.

Question: on this first script I tested only the common longopts used by surface, but there's no reason we couldn't test all of the others as well. This would simplify the cut/paste on future tests. I can't think of a good reason not to do this, but maybe you can?

**Comment by PaulWessel** [comment]
Well, no reason to test —region=whatever in every module that uses -R since it gets repetitive.  Maybe we have one script for each common option and its many incarnations, while other common options (e.g., -r, -j) can team up in a single script. That lets the module test avoid all those -R -J repeated settings.
Ok, bedtime for Bonzo!

**Comment by rbdavis** [comment]
OK, I will make one test script for common opts and then omit those from the per-module scripts. Just let me know where you would like them located in the git test/* structure, e.g., all in one subdir, or each module's test in the same per-module subdir as that module's other tests. (I suppose if you choose the latter we'll still need another special place to stash the common opts test script.)

**Comment by PaulWessel** [comment]
Put the common opts test in test/common (create it) and use the existing dirs for more test scripts for any module.  I will have to figure out how we tell Cmake to do or not do those special tests.  I will do that while you are gone I think when you have at least one or two scripts in common and at least one or two for modules.  Should not be too hard but may involve some special check in the script to see if it should be run or not.

**Comment by rbdavis** [comment]
Will do. Planning to work Mon/Tues this week, in any case will try to post PR by the end of the week including all parsing changes to gmt_init.c and as many test scripts as I can finish. Hopefully that will give enough time for anyone who wants to look things over and have me change anything by the end of the following week.

**Comment by rbdavis** [comment]
Working on **pscoast** test now and encountered a problem: there are two conflicting longopts entries for **--panel** in the common .h file AND the module-specific file. No idea how or when this bug may have popped up (either I just overlooked it or perhaps during the pscoast  PR someone introduced this conflict), in any case **--panel** was never called by the existing pscoast test scripts so failure was non-obvious. It's obvious now, though, so what should we do? Options are basically remove one of these entries or change one of the longopt keywords. The common entry is very skeletal:

`        {   0, 'c', "panel",                 "", "", "", "" },
`

whereas the pscoast-specific entry is more filled out:

[ Edit to original comment, where I mis-copied the wrong entry from the pscoasts .h file -- here is the right one. ]

 ```
       { 0, 'F', "panel",               
                  "l,t",               "scale,rose",
                  "c,g,i,p,r,s",       "clearance,fill,inner,pen,radius,shade" },
```
???

**Comment by rbdavis** [comment]
Oops, copied wrong part of pscoast longopts .h file in that last comment/question, should have been

```
        { 0, 'F', "panel",               
                  "l,t",               "scale,rose",
                  "c,g,i,p,r,s",       "clearance,fill,inner,pen,radius,shade" },
```

**Comment by rbdavis** [comment]
Another pscoast question -- seems like an oversight that we did the **-N** longopt in a very skeletal way:

`        { 0, 'N', "borders",           "", "", "", "" },
`

Should we have perhaps dressed it up with these directives?

```
        { 0, 'N', "borders",               
                  "1,2,3,a",       "national,state,marine,all",
                  "",              "" },
```

There appears to be no restriction that shortopt directive characters must be alphabetic-only. Perhaps we didn't do this because it results in some odd-looking shortopt strings with run-on digits, e.g.:

> % gmt pscoast --l2stranstest --borders=national:3p,red --borders=state:1p,yellow
> --l2stranstest -N13p,red -N21p,yellow

but I'm guessing pscoast would still parse these correctly?

**Comment by rbdavis** [comment]
Just discovered another **--panel** conflict, this time in **pswiggle**'s longopts.h file:

```
        { 0, 'F', "panel",             
                  "",                  "",
                  "c,g,i,p,r,s",       "clearance,fill,inner,pen,radius,shade" },
```

Note this entry is very similar to but slightly different than pscoast's. (This one has no long directives.)

In this case the pre-existing pswiggle test scripts did have some **--panel** options, but they would have run with no errors because no directives or modifiers were specified.

I am wondering if I/we made an assumption somewhere that module-specific entries would override common entries, but that is not how the code works. The commons are searched  first, then the module-specific next, and the first match wins. Maybe we want to reverse this, although that might add some code complexity if there was a module with no module-specific longopts file (highly unlikely, though). Or maybe we just don't want there to ever be conflicting entries between the common and module-specific files?

**Comment by rbdavis** [comment]
Hah, looking on the positive side, these new tests are certainly turning up errors!

**Comment by PaulWessel** [comment]
Hi Roger, sorry forgot to ask (or missed it).  What is the name of the branch you are working on so that I can switch to it and play with it? You need to publish the branch I think.

**Comment by rbdavis** [comment]
I just did a 

> % git push -u origin longoptcolons

which is what I usually do just prior to posting each PR. (Branch name is **longoptcolons** .) These changes include the : changes to the translator and a few tests:

> % git status
> On branch longoptcolons
> Changes to be committed:
>   (use "git restore --staged <file>..." to unstage)
> 	modified:   src/gmt_common_longoptions.h
> 	modified:   src/gmt_init.c
> 	modified:   src/longopt/pscoast_inc.h
> 	new file:   test/common/common-l2s.sh
> 	new file:   test/grdimage/grdimage-l2s.sh
> 	new file:   test/pscoast/pscoast-l2s.sh
> 	new file:   test/pswiggle/pswiggle-l2s.sh
> 	new file:   test/surface/surface-l2s.sh

They also include the proposed modification to the pscoast **--borders** entry I discussed a few comments back.

Note that in the pscoast and pswiggle tests I have commented out translation tests for the problematic **--panel** entries.

**Comment by PaulWessel** [comment]
Thanks, got it.  Will try after dinner...

**Comment by rbdavis** [comment]
Hope it's as tasty as my breakfast! ;->

**Comment by PaulWessel** [comment]
Note that I merged in ~20 commutes from master into longoptionscolons. Probably a good idea that you merge master into your branch before you work on it so that it is up-to-date with master (apart from the branch changes)

**Comment by rbdavis** [comment]
Not surprising, I started the branch right after returning from Oslo. I usually re-download master and reincorporate my changes just before submitting PR. (Which I'm hoping  to do later today.)

**Comment by rbdavis** [comment]
Ouch, I was dreading this -- just encountered a colon problem with mapproject -E (**--ecef**) ellipsoid specification:

> % gmt mapproject --l2stranstest --ecef=Airy:1,2,3
> --l2stranstest -EAiry

By my reading of the mapproject man page the following should be a valid -E option, and mapproject indeed does not seem to complain about it:

> % gmt mapproject -EAiry:1,2,3
> 5 6
> 6318722.09787	552818.464046
> 0 0
> 6377564.396	2

The translator leaves colons untouched only if they are both preceded and followed by digits, e.g., "23:59:59". Options here:

(1) Abandon the **--ecef** longopt.

(2) Abandon single-colon syntax in the translator and revert to ":=" as we were doing before.

(3) Revisit the translator code and try to find a special-case solution that does not unreasonably increase complexity.

???

**Comment by rbdavis** [comment]
Looking into this for a while:

> (3) Revisit the translator code and try to find a special-case solution that does not unreasonably increase complexity.

It may not be hopeless. If I get stumped I'll revert to working on other non-problematic test scripts until we can discuss further.

**Comment by PaulWessel** [comment]
Hard to know how many of these we might find, which sort of determines how to solve it.  I guess hold off with long options for things that breaks and then we revisit when we know.

**Comment by rbdavis** [comment]
OK, so are you saying for now just get rid of the **--ecef** longopt? I guess I can leave it commented out in the .h with a description of the problem.

I'm playing with the debugger right now to see if there might be a fix, probably worth goofing with it a bit longer and if I come up with anything useful I'll let you know.

**Comment by PaulWessel** [comment]
Sure, see what you can do - last resort is commenting and maybe flag the comment (`/* REVISIT_LONG_OPTION */ `or something) so it is easy to **grep** later.
Bedtime for Bonzo.

**Comment by rbdavis** [comment]
Will do, have a good night's sleep! I have an idea to try that could work, but will probably need to do some serious retesting to ensure I haven't broken anything else.

Just a reminder I will likely not be back at this after today until next week, but will check back here before then. If you have any thoughts on the **--panel** and **--borders** questions please advise tomorrow or over the weekend and I will take those up first thing next week, too.

**Comment by rbdavis** [comment]
I think I may have fixed it in the translator:

> % gmt mapproject --l2stranstest --ecef=Airy:1,2,3
> --l2stranstest -EAiry:1,2,3

Colons surrounded by digits are ignored as before. If a colon which otherwise seems valid is located, it too will be ignored if what precedes it is not a longopts string. I ran all my new test scripts with no errors.

I tried to upload the modified gmt_init.c but that failed due to conflicts so I just e-mailed  that to you separately if you want to play with it. Will postpone my planned PR request until next week pending your comments, and continue with test scripting for the rest of today.

**Comment by rbdavis** [comment]
Were there some recent additions made to **mapproject -W** ? The man page has a few **-W** directives (b,B,m,M) that were not in the existing mapproject longopts .h file. The plot in the man page looks suspiciously like stuff that was being discussed in Oslo with Kristof.

I have updated the mapproject longopts .h to match.

**Comment by rbdavis** [comment]
Hi Paul, hope you had a good weekend!

I just submitted the PR for the new longoptions colon stuff and first batch of translation tests. Used a new branch longoptcolons2 to avoid confusion with the first longoptcolons branch I uploaded to git last week so you could test.

Most of the questions I posted here last week remain outstanding in terms of needing your opinion. In the end I left mapproject as-is (i.e., I did not add the new longoptions for what seem like new -W directives) and also did not add the proposed new --borders stuff for pscoast. And the --panels question also remains unresolved.

Will proceed with further translation tests to catch up on all the other longoptified modules.

**Comment by PaulWessel** [comment]
OK, sorry not following up - side-tracked by other things.  As of 1 hour ago we have a power outage, and this week I have a few doctors apt and treatments.

**Comment by rbdavis** [comment]
No worries, hope the medical stuff goes well! I'm at ARL today and tomorrow, will return to GMT Thursday and Friday. I'll post another PR at the end of the week to include whatever I manage to get done by then (probably more translation test scripts) before I leave on my mainland trip.

**Comment by rbdavis** [comment]
Hi Paul, I'm working on the psxy translation test right now and have a question about **-E** (errorbars). Looking in more detail at the manual page, I think we may have a problem with the current longopts .h entry for this:

```
        { 0, 'E', "errorbars",
                  "x,y,X,Y",                 "xbar,ybar,boxwhisker,stemleaf",
                  "a,A,c,n,w,p",             "asymmetrical,lhbounds,symbolfill,notch,capwidth,pen" }
```

The (possible) problem here is that more than one directive can be specified according to the manual page, e.g., **-Exy** or **-EXY**. There's no way that this can be expressed with the current (or any possible, I believe) longopts entry. You can do **--errorbars=xbar** or **--errorbars=ybar** , but you can't do **--errorbars=xbarybar**. I think we need to change that entry to not specify any longopt directives and instead just read

```
        { 0, 'E', "errorbars",
                  "",                               "",
                  "a,A,c,n,w,p",             "asymmetrical,lhbounds,symbolfill,notch,capwidth,pen" }
```

This seems like another case where being able to specify a shortopts string rather than just a shortopts char would be useful, e.g.,

`"x,y,xy,X,Y,XY",                 "xbar,ybar,xybars,boxwhisker,stemleaf,boxandstem",
`

or something like that. Of course, in this case if you can also do **-ExY** and **-EXy** then this gets messier but I suppose is still manageable.

With the short fuse on my 6-week trip, I don't think I should make further translation code enhancements at this moment, although we could work on that when I return. In the meantime maybe we need to get rid of these psxy longopts directives?

**Comment by rbdavis** [comment]
... or maybe there are other workarounds, e.g., either specify multiple -E options (if that's possible, the man page does not say), or force users to use only shortopts -E if they want -Exy ?

**Comment by PaulWessel** [comment]
Yes, I was thinking that too. Since only old scripts might do -Exy we don’t to worry about that one (pass on as is or split in two, whatever makes the parser simplest

**Comment by rbdavis** [comment]
OK, so no change for now to the psxy longopts .h file, correct?

**Comment by PaulWessel** [comment]
Yep

**Comment by rbdavis** [comment]
Having a problem doing the translation test on the **-M** option for **movie** :

**-M**[_frame_|**f**|**m**|**l**],[_format_][**+r**_dpu_][**+v**]

According to this syntax spec it appears that the comma preceding [_format_] is mandatory. Is this correct? If so, then the following translation test fails because that comma is being lumped together with the preceding **first** longopt directive so that the longopt directive string is not recognized:

> % gmt $m $l2s --master=first,png+view
> --l2stranstest -Mfirst,png+v

If that comma really needs to be there then the only quick fix I see here is to get rid of all the **-M** longopt directive strings.

PS: **movie** seems to have some new shortopts ( **-A** ) and short modifiers ( **+v** for **-F** and **-M** ), so I am adding them to its longopts .h file. There seems to be no Julia or Python support for **movie** as far as I can tell.

**Comment by rbdavis** [comment]
Working on **psevents** just now, and noting that we could not do longopt support for its **-Jz|z**_parameters_ option because that conflicts with the common **-J** longopt. Makes me wonder again if we should be searching the module-specific .h before searching the common .h, then we could support things like this. That might introduce other problems somehow and you would know better than I, but something to think about over the next several weeks while I am away. (This also has bearing on the **--panels** questions I asked recently here.)

**Comment by PaulWessel** [comment]
Maybe.  Lots of modules take both -J (horizontal map projection) and -Jz|Z vertical scaling for 3-D. Maybe we need a code section that passes -Jz|Z to one long-option or else the common one.  Maybe would need to keep both in the common.h somehow.

**Comment by rbdavis** [comment]
FYI, found a bug in the **grdinterpolate** .h file -- had to remove the longopts modifier strings for the **-F** option's **+d** modifier. There was an obvious bug in the entry: `{  0, 'F', ... "1,2", "fderiv,sderiv" } ` will not translate **+fderiv** to **+d1**. Probably my own original error.

**Comment by rbdavis** [comment]
Correction on the previous comment from yesterday, I eventually did this:

```
     { 0, 'F', "interptype",
                  "l,a,c,n",               "linear,akima,cubic,none",
                  "d",                       "derivative" },
```

Now at least this works:

> % gmt $m $l2s --interptype=linear+derivative:1
> --l2stranstest -Fl+d1

**Comment by rbdavis** [comment]
Found another rogue option that breaks the translator -- **pspolar -Q** has a **+v** modifier that takes an arg string that can include **+** characters, e.g., from the man page: **+v0.3i+e+gblack** . I have removed **+v** from the longopts .h  file for now, but this raises a general question. Just removing the **+v** does not fix the problem insofar as specification of **-Q** via longopts (i.e., via **--mode** ) will still trigger an error:

> % gmt $m $l2s --mode=spolarity:green+v
> gmt [WARNING]: Long-modifier form v for option -Q not recognized!
> --l2stranstest -Qsgreen

The only way you can get the command to run is to not use the longopts **--mode** version of **-Q** at all:

> %gmt $m $l2s -Qsgreen+v0.3i+e+gblack
> --l2stranstest -Qsgreen+v0.3i+e+gblack

Now, we could actually leave **+v** in the longopts .h and it would work when called with **--mode** as long as no **+** characters were part of the **+v** argument string. If we do not have **+v** in the longopts .h it will never work when called with **--mode** whether or not **+** characters are present. So, should we restore **+v** to the longopts .h and allow users to successfully use it when no extra **+** characters are there, since they can otherwise never use **+v** at all when calling via **--mode** if **+v** is removed from the longopts .h?

**Comment by PaulWessel** [comment]
Yes, what a stupid decision I made back then, but I did it to fit with the parsing at that time. +v passes on arguments that are given as modifier as well, and that was bad.   I think flag it and not attempt.  I am guessing I would prefer to "modernise" that syntax so that the l2s works on that flavor (whatever it will be) but otherwise we must either pre-process the +v+e+g mess into the new format and then the l2s will work.

**Comment by rbdavis** [comment]
OK, just did another PR including a large batch of l2s translation tests. I did not quite catch up to the backlog of existing longoptified modules -- there are still 6 modules without tests yet. Remember to approve the outstanding PR I submitted a while back before doing this one.

I'll be back at it when I return at the end of September, and hope all is well with you in the meantime! (I'll be on e-mail via phone only, cannot do any useful work remotely as I'm travelling sans laptop.)
